### PR TITLE
ignore hashtag suggestions if they vary only in case

### LINKED ISSUE #16449: Hashtag suggestions are pointlessly replacing CamelCase hashtags
### Expected behaviour

When typing a hashtag in CamelCase, it should remain in CamelCase (as this is important for readability, especially for those using screen readers)

### Actual behaviour

If the CamelCase hashtag matches an existing hashtag, the auto-suggest feature is usually removing the CamelCase and making it all lower case

### Steps to reproduce the problem

1. Try to type a popular hashtag in CamelCase (for example DogsOfMastodon) 
2. Autosuggest box will appear, and will (probably) suggest the same tag but entirely lower case (dogsofmastodon)
3. When you press enter to get rid of the suggestion box, the hashtag has the CamelCase stripped out
...

This seems to be caused by suggestions being taken from whatever tags were first used on that instance. If an entirely lower case tag was used first, that is what ends up being suggested.

It would help a lot if the suggestion system simply did not replace a hashtag if the original text string and suggested text string is an exact match. This would (hopefully) leave the CamelCase as inputted by the user intact.

### Specifications

Mastodon v3.4.1
Firefox, Ubuntu

### PR DESCRIPTION
**Pull request by weex**
Fixes #16449 by adding an ignoreSuggestion event.

### CODE REVIEW

File: app/javascript/mastodon/reducers/compose.js [hunk 1]
```diff
@@ -155,6 +156,16 @@ const insertSuggestion = (state, position, token, completion, path) => {
   });
 };
 
+const ignoreSuggestion = (state, position, token) => {
+  return state.withMutations(map => {
+    map.set('suggestion_token', null);
+    map.set('suggestions', ImmutableList());
+    map.set('focusDate', new Date());
+    map.set('caretPosition', position + token.length + 1);
```
Comment (ClearlyClaire) [inline app/javascript/mastodon/reducers/compose.js:164]: This introduces a slight behavior inconsistency, with a space being added if the hashtag isn't a case-insensitive match, and no space being added if there is.

**Review by ClearlyClaire — CHANGES_REQUESTED** [review]
I wonder if, instead of adding an event type, it wouldn't be better to change `insertSuggestion` so that it selects either the `completion` or `token` based on the case-insensitive comparison.

I also wonder if it wouldn't make sense, if the two strings differ in casing only, and one is lowercase-only, to select the other, as it is most likely more readable.

**Comment by weex** [comment]
> I wonder if, instead of adding an event type, it wouldn't be better to change `insertSuggestion` so that it selects either the `completion` or `token` based on the case-insensitive comparison.

The issue there is that the token is already lowered when populating textAtCursorMatchesToken (https://github.com/tootsuite/mastodon/blob/main/app/javascript/mastodon/components/autosuggest_textarea.js#L11) so either we need to pass through another argument or return the non-lowered token there.

> I also wonder if it wouldn't make sense, if the two strings differ in casing only, and one is lowercase-only, to select the other, as it is most likely more readable.

It might but it's probably better in another PR.

File: app/javascript/mastodon/reducers/compose.js (hunk 1 above)
Comment (weex) [reply app/javascript/mastodon/reducers/compose.js:164]: Great catch. I'm not sure how to best to generate a space of it is necessary.

**Comment by ClearlyClaire** [comment]
> > I wonder if, instead of adding an event type, it wouldn't be better to change `insertSuggestion` so that it selects either the `completion` or `token` based on the case-insensitive comparison.
> 
> The issue there is that the token is already lowered when populating textAtCursorMatchesToken (https://github.com/tootsuite/mastodon/blob/main/app/javascript/mastodon/components/autosuggest_textarea.js#L11) so either we need to pass through another argument or return the non-lowered token there.

Ah, I missed the `token` being lowercased already. It's a bit convoluted, but the initial text can easily be recovered by doing something like `map.get(path).slice(position, position + token.length)`.

> > I also wonder if it wouldn't make sense, if the two strings differ in casing only, and one is lowercase-only, to select the other, as it is most likely more readable.
> 
> It might but it's probably better in another PR.

File: app/javascript/mastodon/reducers/compose.js (hunk 1 above)
Comment (ClearlyClaire) [reply app/javascript/mastodon/reducers/compose.js:164]: Something like (untested):
```javascript
map.updateIn(path, oldText => `${oldText.slice(0, position + token.length)} ${oldText.slice(position + token.length)}`);
```

**Comment by rkingett** [comment]
This is a welcomed proposed change.

File: app/javascript/mastodon/reducers/compose.js (hunk 1 above)
Comment (weex) [reply app/javascript/mastodon/reducers/compose.js:165]: Added that which worked.

**Comment by weex** [comment]
Thanks @ClearlyClaire, I updated with your proposed change which cleared the inconsistency about the space.

This PR isn't perfect but it's progress so leaving this issue for now. The comments above about `map.get(path).slice(position, position + token.length)` might be a better way to solve if anyone else wants to tackle.

**Review by ClearlyClaire — APPROVED** [review]
I still prefer my earlier suggestion, but this PR looks good to me and is an improvement upon the current behavior.

**Comment by ClearlyClaire** [comment]
Was this closed by error?

**Comment by weex** [comment]
Yes

**Comment by FediVideos** [comment]
Is there some reason this was closed?

**Comment by weex** [comment]
Thanks for the ping. This was created when the magicstone-dev repo was my personal one and since then I've been trying to delete any branches other than `main` due to 2.5.1 [here](https://rfc.zeromq.org/spec/42/#25-branches-and-releases). So as you see from the history I've deleted and recreated the branch a couple of times, while reopening the PR is less obvious a step after that.

**Comment by weex** [comment]
Can't tell what the problem was in CI since I suppose the logs have been deleted after a few months in DeepSource. If this gets re-run, please ping me to let me know if anything needs to be done.

**Comment by ClearlyClaire** [comment]
It likely was an issue with DeepSource itself, and not your code (sometimes, it would basically re-check everything, not only the changed parts). I'd rebase and force-push.

**Review by Gargron — APPROVED** [review]

**Comment by FediVideos** [comment]
Has this change been implemented in the latest version of Masto? At the moment I get the following results:

On Mastodon.online (v3.4.6): Try to type #HashTag but suggestions turn it into #häshtäg (with umlauts!)

On Mas.to (v3.4.6): Try to type #HashTag but suggestions turn it into #hashtag

On Mstdn.social (v3.4.6+glitch): Try to type #HashTag, successfully stays as #HashTag

**Comment by FediVideos** [comment]
So, it is there but broken on mastodon.online? Just tried it again and it still does that umlaut thing...

Is there a correct procedure for asking this to be fixed? (Sorry, I'm not that familiar with how github works!)

**Comment by rkingett** [comment]
Just checked on my instance, writing exchange, and it is still changing it to all lower case. I wish the opposite would happen, where all hashtags were changed to CammelCase. In order for this to work, this PR must be merged into the main branch, which it looks like it has?

**Comment by weex** [comment]
Tried it on fosstodon (v.3.4.6_ and it seems to be working as expected. @FediVideos the best place is probably to create a new issue. Could be a subtle way in how you're using it? Timing of hitting tab vs. enter vs. space. In a new issue you could describe exactly what you're doing and mention this issue number so there's a trail for later peeps to follow.

**Comment by ClearlyClaire** [comment]
Note that this has *not* landed in any release, it's only in the development version (which is currently also reported as 3.4.6).

**Comment by FediVideos** [comment]
Weird, it is now working perfectly on mastodon.online!

I am typing HashTag, the top suggestion is hashtag but pressing space leaves me with HashTag (which is exactly how it should behave!).

So, when is this expected to roll out to release versions?

**Pull request merged**